watertap/property_models/SOP_prop_pack.py

```diff
@@ -0,0 +1,551 @@
+            if number_unfixed_variables(self[k]) != 0:
+                skip_solve = False
+
+        if not skip_solve:
+            # Initialize properties
+            with idaeslog.solver_log(solve_log, idaeslog.DEBUG) as slc:
+                results = solve_indexed_blocks(opt, [self], tee=slc.tee)
+            init_log.info_high(
+                "Property initialization: {}.".format(idaeslog.condition(results))
+            )
+
+        # ---------------------------------------------------------------------
+        # If input block, return flags, else release state
+        if state_vars_fixed is False:
+            if hold_state is True:
+                return flags
+            else:
+                self.release_state(flags)
+
+    def release_state(self, flags, outlvl=idaeslog.NOTSET):
+        """
+        Method to release state variables fixed during initialisation.
+
+        Keyword Arguments:
+            flags : dict containing information of which state variables
+                    were fixed during initialization, and should now be
+                    unfixed. This dict is returned by initialize if
+                    hold_state=True.
+            outlvl : sets output level of of logging
+        """
+        # Unfix state variables
+        init_log = idaeslog.getInitLogger(self.name, outlvl, tag="properties")
+        revert_state_vars(self, flags)
+        init_log.info_high("{} State Released.".format(self.name))
+
+
+# this third class, provides the instructions to create state blocks
+@declare_process_block_class("SopStateBlock", block_class=_SopStateBlock)
+class SopStateBlockData(StateBlockData):
+    def build(self):
+        """Callable method for Block construction."""
+        super(SopStateBlockData, self).build()
+
+        self.scaling_factor = Suffix(direction=Suffix.EXPORT)
+
+        # first we create the state variables
+        self.flow_mass_phase_comp = Var(
+            self.params.phase_list,
+            self.params.component_list,
+            initialize={("Liq", "H2O"): 1, ("Liq", "oil"): 0.01},
+            bounds=(1e-15, 100),
+            domain=NonNegativeReals,
+            units=units.kg / units.s,
+            doc="Mass flow rate",
+        )
+
+        self.temperature = Var(
+            initialize=298.15,
+            bounds=(273.15, 1000),
+            domain=NonNegativeReals,
+            units=units.degK,
+            doc="State temperature",
+        )
+
+        self.pressure = Var(
+            initialize=101325,
+            bounds=(1e5, 5e7),
+            domain=NonNegativeReals,
+            units=units.Pa,
+            doc="State pressure",
+        )
+
+    # -----------------------------------------------------------------------------
+    # Property Methods
+    # these property methods build variables and constraints on demand (i.e. only when asked
+    # for by a user or unit model)
+    def _visc_d_phase_comp(self):
+        self.visc_d_phase_comp = Param(
```

Comment:
I suppose we could get away with this, but for consistency purposes, we typically build the params on the parameter block above, then here, we'd do an `add_object_reference`. You can check out the multi_aq_sol_prop_pack and see an example or two of that (check it out for viscosity since the case for diffusivity may be a bit more complicated to read through)